### PR TITLE
Canonize and cast to

### PR DESCRIPTION
**Pull request by BilelGho**
Implementing canonize and cast to

### CODE REVIEW

File: src/cdomains/intDomain.ml [hunk 1]
```diff
@@ -942,18 +942,47 @@ struct
...
 
   let show (x: t) =
     let show_interval i = Printf.sprintf "[%s, %s]" (Ints_t.to_string (fst i)) (Ints_t.to_string (snd i)) in
     List.fold_left (fun acc i -> (show_interval i) :: acc) [] x |> String.concat ", " |> Printf.sprintf "[%s]"
   
-
-  let canonize x = failwith "Not implemented yet"
+  type 'a event = Enter of 'a | Exit of 'a 
+  let unbox = function Enter x -> x | Exit x -> x
+  let operand_to_events (xs:t) =  List.map (fun (a,b) -> [Enter a; Exit b]) xs |> List.flatten
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:953]: interval_set_to_events

File: src/cdomains/intDomain.ml [hunk 2]
```diff
@@ -942,18 +942,47 @@ struct
...
+  
+  let operands_to_events (xs:t) (ys:t) = (xs @ ys)  |> operand_to_events |> List.sort (fun x y -> Ints_t.compare (unbox x) (unbox y))
+    
+  let combined_event_list (xs: int_t event list) lattice_op =
+    let l = match lattice_op with `Join -> 1 | `Meet -> 2 in
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:958]: add a comment explaining what is happening ( why join is 1 and meet is 2)

File: src/cdomains/intDomain.ml [hunk 3]
```diff
@@ -942,18 +942,47 @@ struct
...
+      let aux (interval_count,acc) = function
+      | Enter x -> (interval_count+1, if interval_count+1<= l && interval_count< l then (Enter x)::acc else acc)
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:960]: >=

File: src/cdomains/intDomain.ml [hunk 4]
```diff
@@ -942,18 +942,47 @@ struct
...
+    | _ -> (l,r)::acc
+    in 
+      List.fold_left f [] xs |> List.rev 
+    
+  (* Helper Functions *)
+  let min_list l = List.fold_left min (List.hd l)
+  let max_list l = List.fold_left max (List.hd l)
+  let list_of_tuple2 (x, y) = [x ; y] 
+
+  let canonize (xs:t) = operand_to_events xs |> fun x -> combined_event_list x `Join |> events_to_intervals
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:981]: add remove gaps

File: src/cdomains/intDomain.ml [hunk 5]
```diff
@@ -1109,11 +1113,53 @@ struct
...
 
-  let narrow _x _y _z  = failwith "Not implemented yet"
+  let rec partition_pair (ik :ikind) (acc : (int_t*int_t) option ) (xs:t) (ys:t)= 
+      match xs,ys with 
+      | _, [] -> []
+      |[], (y::ys) -> (acc,y):: partition_pair ik None [] ys 
+      |(x::xs), (y::ys) when  leq_interval (Some x) (Some y) -> partition_pair ik (join_interval ik acc (Some x)) xs (y::ys)
+      |(x::xs), (y::ys) -> (acc,y) :: partition_pair ik None  (x::xs) ys
+  
+  let approaching_pair  x y = match x,y with
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:1137]: partition_pair_is_approaching

File: src/cdomains/intDomain.ml [hunk 6]
```diff
@@ -1109,11 +1113,53 @@ struct
...
+      | _, [] -> []
+      |[], (y::ys) -> (acc,y):: partition_pair ik None [] ys 
+      |(x::xs), (y::ys) when  leq_interval (Some x) (Some y) -> partition_pair ik (join_interval ik acc (Some x)) xs (y::ys)
+      |(x::xs), (y::ys) -> (acc,y) :: partition_pair ik None  (x::xs) ys
+  
+  let approaching_pair  x y = match x,y with 
+   (Some (_,ar),(_,br)),(Some (al,_),(bl,_)) -> Ints_t.compare (Ints_t.sub al ar) (Ints_t.sub bl br) > 0  
+    | _,_ -> true
 
-  let widen _x  = failwith "Not implemented yet"
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:1116]: implement widen

File: src/cdomains/intDomain.ml [hunk 7]
```diff
@@ -1125,20 +1171,75 @@ struct
...
-  let project _ik _ip _x = failwith "Not implemented yet"
-  
-  let arbitrary _ik = failwith "Not implemented yet"
+  let excl_to_intervalset (ik : ikind) ((rl,rh): (int64 * int64)) (excl : int_t): t = 
+    let intv1 = norm ik @@ Some (Ints_t.of_bigint (Size.min_from_bit_range rl), Ints_t.sub excl Ints_t.one) in
+    let intv2 = norm ik @@ Some (Ints_t.add excl Ints_t.one,  Ints_t.of_bigint (Size.max_from_bit_range rh)) in
+   [intv1; intv2] |> List.filter_map (fun x -> x)
+
+  let refine_with_excl_list ik (intv : t) = function
+  | None -> []
```
Comment (BilelGho) [inline src/cdomains/intDomain.ml:1228]: check if this is really the bottom

**Review by gabryon99 — APPROVED** [review]

**Pull request merged**